test/ServiceTest.cpp

```diff
@@ -0,0 +1,179 @@
+#include "./IndexTestHelpers.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+#include "ctre/ctre.h"
+#include "engine/Service.h"
+#include "parser/GraphPatternOperation.h"
+#include "util/http/HttpUtils.h"
+
+// Fixture that sets up a test index and a factory for producing mocks for the
+// `getTsvFunction` needed by the `Service` operation.
+class ServiceTest : public ::testing::Test {
+ protected:
+  // Query execution context (with small test index) and allocator for testing,
+  // see `IndexTestHelpers.h`. Note that `getQec` returns a pointer to a static
+  // `QueryExecutionContext`, so no need to ever delete `testQec`.
+  QueryExecutionContext* testQec = ad_utility::testing::getQec();
+  ad_utility::AllocatorWithLimit<Id> testAllocator =
+      ad_utility::testing::makeAllocator();
+
+  // Factory for generating mocks of the `sendHttpOrHttpsRequest` function that
+  // is used by default by a `Service` operation (see the constructor in
+  // `Service.h`). Each mock does the following:
+  //
+  // 1. It tests that the request method is POST, the content-type header is
+  //    `application/sparql-query`, and the accept header is
+  //    `text/tab-separated-values` (our `Service` always does this).
+  //
+  // 2. It tests that the host and port are as expected.
+  //
+  // 3. It tests that the post data is as expected.
+  //
+  // 4. It returns the specified TSV.
+  //
+  // NOTE: In a previous version of this test, we set up an actual test server.
+  // The code can be found in the history of this PR.
+  static auto constexpr getTsvFunctionFactory =
+      [](const std::string& expectedUrl, const std::string& expectedSparqlQuery,
+         const std::string& predefinedResult) -> Service::GetTsvFunction {
+    return [=](ad_utility::httpUtils::Url url,
+               const boost::beast::http::verb& method,
+               std::string_view postData, std::string_view contentTypeHeader,
+               std::string_view acceptHeader) -> std::istringstream {
+      // Check that the request parameters are as expected.
+      //
+      // NOTE: The first three are hard-coded in `Service::computeResult`, but
+      // the host and port of the endpoint are derived from the IRI, so the last
+      // two checks are non-trivial.
+      EXPECT_EQ(method, http::verb::post);
+      EXPECT_EQ(contentTypeHeader, "application/sparql-query");
+      EXPECT_EQ(acceptHeader, "text/tab-separated-values");
+      EXPECT_EQ(url.asString(), expectedUrl);
+
+      // Check that the (whitespace-normalized) post data contains the expected
+      // query.
+      //
+      // NOTE: a SERVICE clause specifies only the body of a SPARQL query, from
+      // which `Service::computeResult` has to construct a full SPARQL query by
+      // adding `SELECT ... WHERE`, so this checks something non-trivial.
+      EXPECT_EQ(
+          std::regex_replace(std::string{postData}, std::regex{"\\s+"}, " "),
```

Comment:
If you bind this to a variable `whitespaceNormalizedPostData`, 
then the code becomes self-documenting:)